python/cuml/cuml/experimental/accel/estimator_proxy.py

```diff
@@ -0,0 +1,146 @@
+#
+# Copyright (c) 2024, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+
+import cuml
+import importlib
+import inspect
+import os
+
+from cuml.internals.global_settings import GlobalSettings
+from cuml.internals.mem_type import MemoryType
+from cuml.internals import logger
+from cuml.internals.safe_imports import gpu_only_import, cpu_only_import
+from typing import Optional
+
+
+# currently we just use this dictionary for debugging purposes
+patched_classes = {}
+
+
+def intercept(
+    original_module: str,
+    accelerated_module: str,
+    original_class_name: str,
+    accelerated_class_name: Optional[str] = None,
+):
+
+    if accelerated_class_name is None:
+        accelerated_class_name = original_class_name
+    # Import the original host module and cuML
+    module_a = cpu_only_import(original_module)
+    module_b = gpu_only_import(accelerated_module)
+
+    # Store a reference to the original (CPU) class
+    if original_class_name in patched_classes:
+        original_class_a = patched_classes[original_class_name]
+    else:
+        original_class_a = getattr(module_a, original_class_name)
+        patched_classes[original_class_name] = original_class_a
+
+    # Get the class from cuML so ProxyEstimator inherits from it
+    class_b = getattr(module_b, accelerated_class_name)
+
+    # todo (dgd): add environment variable to disable this
+    class ProxyEstimatorMeta(cuml.internals.base_helpers.BaseMetaClass):
+        def __repr__(cls):
+            return repr(original_class_a)
+
+    class ProxyEstimator(class_b, metaclass=ProxyEstimatorMeta):
+        def __init__(self, *args, **kwargs):
+            self._cpu_model_class = (
+                original_class_a  # Store a reference to the original class
+            )
+            # print("HYPPPPP")
+            kwargs, self._gpuaccel = self._hyperparam_translator(**kwargs)
```

Comment:
We need  to handle things like `KMeans(8)` as well. So not just translating keyword arguments but also the (few?) occasions where positional arguments are allowed for the scikit-learn estimator.

Right now `kwargs = {'args': 8}` when you instantiate `KMeans(8)` like this. That seems definitely not what we want :D

Reply:
I think we can use the result of `inspect.signature(self._cpu_model_class).bind(*args, **kwargs).arguments` to get a dictionary that contains everything the user passed. 

To get all arguments, including defaults that the user didn't pass:
```python
signature = inspect.signature(self._cpu_model_class).bind(*args, **kwargs)
signature.apply_defaults()
print(signature.arguments)
```
I think this is what we need to pass to the translator

Reply:
I also thought that positional arguments are no longer supported but `KMeans(8) == `KMeans(n_clusters=8)` :-/ I've not looked through the docs to find out which estimators still allow positional arguments, was thinking it would be simpler to "make it work" than try to find out how big a problem it was (given there is [at least one case](https://xkcd.com/1172/)).